### PR TITLE
MCC-2291 Readme updates

### PR DESCRIPTION
**Pull request by jcape**
Soundtrack of this PR: [link to song that really fits the mood of this PR]()

### Motivation

Based on some feedback, the documentation around key derivation, attestation, and the fog signature process needed to be improved. 

### In this PR
* Rewrite the attest README to include a discussion of SGX enclave concepts.
* Rewrite the key derivation README to cover the steps and provide an example usage.
* Rewrite the fog signature documentation to describe the chain of trust created from the user account to fog server.

### CODE REVIEW

File: attest/README.md [hunk 1]
```diff
@@ -1,3 +1,108 @@
...
+1. IAS API call (a REST API call to IAS to sign off on the local attestation data)
+1. Verification (verifying the result from IAS is what we want)
+
+### Between Enclaves on the same CPU
+
+> Don't skip this section just because you're interested in remote attestation. It's a core building block if you want to understand how the system works.
+
+The first thing to know about attestation is that it begins with Alice The Enclave asking the CPU for it's own `EREPORT`. The CPU will fill in a data structure containing information about Alice.
+
+The contents of Alice's `EREPORT` can be used to construct a `TARGETINFO` data structure, which is given to Bob The Enclave. Bob Alice's targetting information when he calls `EREPORT` from within his context, and he CPU fills in the report as before, and adds an authentication code so Alice can verify Bob's report is valid.
```
Comment (eranrund) [inline attest/README.md:64]: Bob Alice's?

File: attest/README.md [hunk 2]
```diff
@@ -1,3 +1,108 @@
...
+
+Bob sends his `EREPORT` (with authentication code) to Alice, and she verifies the code using information available only to here and the CPU, and she can trust the report that she got from Bob is actually from Bob. That is, she's authenticated the report, and knows there was a Bob enclave out there, and she knows she has a report for it.
```
Comment (eranrund) [inline attest/README.md:66]:
```suggestion
Bob sends his `EREPORT` (with authentication code) to Alice, and she verifies the code using information available only to her and the CPU, and she can trust the report that she got from Bob is actually from Bob. That is, she's authenticated the report, and knows there was a Bob enclave out there, and she knows she has a report for it.
```

File: attest/README.md [hunk 3]
```diff
@@ -1,3 +1,108 @@
...
+
+The QE has (on startup) established a secure connection to Intel and gotten itself "provisioned" with a key it can use for communicating back to IAS. When it gets the report from Bob, the QE first authenticates the report (as Alice did above), then fills in extra details about the host: it's own version and details, the BIOS version, some security-relevant BIOS configuration details (e.g. "is hyperthreading turned on" or "is the GPU turned on", etc.), etc. and creates what's called a `Quote`.
+
+In addition to the contents of Bob's `Quote` is the contents of Bob's `EREPORT`, details about the QE and the host, and additional information which is encrypted between the QE to Intel.
+
+### IAS
+
+One the MobileCoin application has gotten a `Quote` back from the QE, it encodes it into a JSON document and sends it to the *Intel Attestation Service* (IAS). IAS, for it's part, looks at the quote and tries to verify everything: is the QE up-to-date? Does the CPU have any known vulnerabilities? IAS, then, is verifying that there aren't any known problems with the TCB---that you can actually T the CB.
```
Comment (eranrund) [inline attest/README.md:82]:
```suggestion
Once the MobileCoin application has gotten a `Quote` back from the QE, it encodes it into a JSON document and sends it to the *Intel Attestation Service* (IAS). IAS, for it's part, looks at the quote and tries to verify everything: is the QE up-to-date? Does the CPU have any known vulnerabilities? IAS, then, is verifying that there aren't any known problems with the TCB---that you can actually T the CB.
```

File: attest/README.md [hunk 4]
```diff
@@ -1,3 +1,108 @@
...
+
+It puts all that (alongside "update info" if something is not up-to-date) into a report, and signs the bytes using an RSA key, then provides a standard certificate chain.
+
+### Verification
+
+The software the MobileCoin enclave is attesting to (the client) must now verify the report provided by IAS. In particular, it must authenticate the message was signed by a key it trusts. Doing so is fairly straightforward RSA signature checking, where the verifier is given an X509 certificate chain and signature alongside the  authenticate that the chain leads back to a trust anchor the application expects.
```
Comment (eranrund) [inline attest/README.md:88]:
```suggestion
The software the MobileCoin enclave is attesting to (the client) must now verify the report provided by IAS. In particular, it must authenticate the message was signed by a key it trusts. Doing so is fairly straightforward RSA signature checking, where the verifier is given an X509 certificate chain and signature alongside the authenticate that the chain leads back to a trust anchor the application expects.
```

**Review by eranrund — DISMISSED** [review]
Nice!

File: fog/sig/README.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+## This Crate
+
+This crate provides a `Verifier` trait and an implementation for [`PublicAddress`](mc_account_keys::PublicAddress). This trait contains a single method, `verify_fog_sig()`, which is used to verify a user has signed off on a list of Attestation Verfiication Reports for ingest enclaves.
+
+## In Context
+
+Fog works by having a sender encrypt a transaction's destination public address using a cryptobox-like scheme based on Ristretto255 and AES-256-GCM. The sender looks up a destination user's fog server, determines the appropriate ingest enclave's Ristretto public key, and encrypts the recipient's public address for that enclave.
+
+This family of crates provides primatives to authenticate that a particular user has delegated fog operations (via a chain of trust) to a list of particular ingest server public keys.
```
Comment (UkoeHB) [inline fog/sig/README.md:11]: #### proposed clarification

This family of crates provides primitives to authenticate that a particular user has delegated fog operations to a list of particular ingest server public keys. Specifically, that the user signed off on a Fog operator's root certificate, that there is a certificate chain between that root and some ingest server AVRs, and that those AVRs are valid.

File: fog/sig/README.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+| Crate | Purpose |
+| ----- | ------- |
+| `mc-fog-sig-authority` | Signatures over DER-encoded `subjectPublicKeyInfo` bytes |
+| `mc-fog-sig-report` | Signatures over [`Report[]`](mc_fog_api::Report) structures |
+| `mc-fog-sig` | One-shot validation of [`ReportRequest`](mc_fog_api::ReportRequest) using a [`PublicAddress`](mc_account_keys::PublicAddress)
+
+## Scheme Details
+
+What we want to do is provide a way for the recipient to that they trust a particular fog operator (so far as they need to), and for that fog operator to provide a list of ingest Attestation Verification Reports (from Intel's IAS service), and for a sender to be able to verify this.
```
Comment (UkoeHB) [inline fog/sig/README.md:25]: >  for the recipient to that they trust

for the recipient to trust

File: fog/sig/README.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+### Recipients
+
+Recipient applications perform the following steps for their users:
+
+1. Using the private cryptonote view key (a Ristretto255 scalar), creates a digital signature over a long-lived X.509 certificate authority controlled by the Fog operator.
+1. Publishes a [`PublicAddress`](mc_account_keys::PublicAddress), which contains containing the "Fog URL" (where to query the operator for reports), the "Fog Signature" (raw [`Signature`](schnorrkel::Signature) bytes), and a report ID.
```
Comment (UkoeHB) [inline fog/sig/README.md:32]: > contains containing

contains

> and a report ID.

clarification: and a fog report ID.

**Review by UkoeHB — CHANGES_REQUESTED** [review]

File: attest/README.md [hunk 8]
```diff
@@ -1,3 +1,108 @@
...
+
+### Between Enclaves on the same CPU
+
+> Don't skip this section just because you're interested in remote attestation. It's a core building block if you want to understand how the system works.
+
+The first thing to know about attestation is that it begins with Alice The Enclave asking the CPU for it's own `EREPORT`. The CPU will fill in a data structure containing information about Alice.
+
+The contents of Alice's `EREPORT` can be used to construct a `TARGETINFO` data structure, which is given to Bob The Enclave. Bob uses Alice's targetting information when he calls `EREPORT` from within his context, and he CPU fills in the report as before, and adds an authentication code so Alice can verify Bob's report is valid.
+
+Bob sends his `EREPORT` (with authentication code) to Alice, and she verifies the code using information available only to her and the CPU, and she can trust the report that she got from Bob is actually from Bob. That is, she's authenticated the report, and knows there was a Bob enclave out there, and she knows she has a report for it.
```
[the remark this answers is not in the document]
Comment (jcape) [reply attest/README.md:66]: I agree with EREPORT (the command) vs. REPORT (the structure), but I'm going to keep it simple, there's no need to discuss EGETKEY (which is never directly called in any of our code).

File: fog/sig/README.md (hunk 7 above)
Comment (jcape) [reply fog/sig/README.md:32]: There is no other type of report ID.

File: attest/README.md [hunk 9]
```diff
@@ -1,3 +1,108 @@
...
+
+Bob and Alice now have a secret which each of them knows, but Eve cannot know, which means they can use this shared secret to create their own message authentication codes, between each other, without relying on special CPU instructions like `EREPORT`. When Bob wants to send a message to Alice, he can compute the authentication code for it (the MAC), and send both of them to Alice. Alice, for her part, will independently compute the MAC based on the message, and then compare her MAC with the one Bob sent over. If they match, then she knows that Eve didn't change the message.
+
+In other words, Alice can now authenticate messages from Bob, and visa-versa. If Alice and Bob also want to *hide* their messages from Eve, then Bob would first encrypt his message, then compute the MAC of the encrypted text. It's usually easier to both encrypt and MAC, because there is a special kind of encryption "Authenticated Encryption" that combines encryption and authentication codes into a single package.
```
[the remark this answers is not in the document]
Comment (jcape) [reply attest/README.md:108]: I think this might be better served by a separate section explaining that client-to-node communications work in basically the same way.

File: attest/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (UkoeHB) [reply attest/README.md:66]: Ok, but I think the 'message' concept is important. Attestation is meaningless without the report message.

File: attest/README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (UkoeHB) [reply attest/README.md:108]: This section felt confusing because enclaves on the same machine aren't opening a communication channel with each other in practice. Better to reference the most 'up-to-date' concepts in the readers' minds - AVRs and client-to-enclave.

File: fog/sig/README.md (hunk 7 above)
Comment (UkoeHB) [reply fog/sig/README.md:32]: The reader doesn't know that, necessarily.

File: attest/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (jcape) [reply attest/README.md:66]: I disagree, not because `REPORTDATA` isn't the message (in a cryptographic sense), but because it should not be treated as a "message" by an author of an application that wants to use an enclave. It may be interesting to know, but it isn't necessary to understanding, and will lead folks down the wrong path.

For example: I think one mistake we made was putting the block signing key into the AVR. Now, if we want to save round-trips by adding more details to authentication (e.g. adding the configured minimum fee to that would save a round-trip during tx construction, and an enclave verifier could check that the fees match directly, and provide better error reporting, rather than stuffing it into the noise prologue), we're kind of screwed. If we'd treated those bytes as an authentication tag for a larger data structure instead (i.e. as a bootstrap for a *real* message), I think we'd be in a better spot.

[the remark this answers is not in the document]
Comment (UkoeHB) [reply attest/README.md:66]: I don't really understand. Isn't proving a piece of data (message) was made by an enclave the entire raison d'etre of attestation? You don't gain anything from a `REPORT` or AVR that has no message.

File: attest/README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (jcape) [reply attest/README.md:108]: OK, I'll take a stab at making this more clear.

**Comment by remoun** [comment]
This still seems useful. Do we still want to merge it?

File: account-keys/slip10/README.md [hunk 10]
```diff
@@ -1,5 +1,46 @@
...
+
+// FIXME: Let the user backup their phrase
+
+// Create a SLIP-0010 private child key for account "1"
+let slip10 = mnemo.derive_slip10_key(1)
+
+// Derive the AccountKey from the Slip10Key
+let account_key = AccountKey::from(slip10);
+
+// go forth and spend money!
```
Comment (iamalwaysuncomfortable) [inline account-keys/slip10/README.md:31]:
```suggestion
// you now control an account which can send and receive mobilecoin
```

File: account-keys/slip10/README.md [hunk 11]
```diff
@@ -1,5 +1,46 @@
...
+```
+
```
Comment (iamalwaysuncomfortable) [inline account-keys/slip10/README.md:33]:
```suggestion
Note: because all accounts are Ristretto255 points, accounts do not have to be "registered" with the Mobilecoin network. Once keys are generated, users in possession of those keys can immediately receive or spend Mobilecoin.

```

File: attest/README.md [hunk 12]
```diff
@@ -1,3 +1,108 @@
...
+| `trusted` | Traits used to monkey-patch enclave-only functionality onto `core` types |
+| `untrusted` | Traits used to monkey-patch untrusted-only functionality onto `core` types |
+
+# Enclave Concepts and Jargon
+
+Intel's enclaves and EPID-based remote attestation is a very complex subject (there are many layers and protocols and ways to stitch things together), so the first thing to get out of the way is understanding the basics and the jargon. We'll use the MobileCoin validator (the "consensus service") for this example.
+
+## Trusted Computing Base
+
+***Trusted Computing Base*** (TCB) is pure jargon. It describes the much more pragmatic "all the things you need to believe haven't been pwned (things you trust) in order to believe the system hasn't been pwned (can be trusted)." It naturally follows that one great way to improve security is to reduce the number of things that you have to believe in to be however secure you need to be.
```
Comment (iamalwaysuncomfortable) [inline attest/README.md:21]:
```suggestion
***Trusted Computing Base*** (TCB) is jargon for "all the things you need to trust that haven't been compromised in order to believe the system can be trusted."
```

File: attest/README.md [hunk 13]
```diff
@@ -1,3 +1,108 @@
...
+
+To show how a TCB comes into being, lets take your typical "take a JSON document from HTTP, process it, return more JSON" web application, for example, isn't going to be passing raw network card buffers in RAM around from function to function. Instead you're going to have the OS deserialize the ethernet frame (and IP packet, and TCP segment), a TLS library decrypt and verify the encrypted data, a web server deserialize the HTTP request, and a framework/parser deserialize the JSON into an object, all before your application ever sees anything.
```
Comment (iamalwaysuncomfortable) [inline attest/README.md:23]:
```suggestion
To show how a TCB comes into being, take a typical scenario of a web application receiving a JSON document via HTTP, processing it and returning a JSON response. Encapsulation makes this seem like a single process, but under the hood, the OS deserializes the ethernet frame/IP packet/TCP segment, a TLS library decrypts and verifies the encrypted data, a web server deserializes the HTTP request, and a framework/parser deserializes the JSON into an object before the application sees the original message.
```

File: attest/README.md [hunk 14]
```diff
@@ -1,3 +1,108 @@
...
+
+And ultimately, you're going to create a TCB simply by following good development practices. Stabilizing a procedure and including it in a library that is shared in multiple places makes that library part of your TCB. You've established an ad-hoc trust in that procedure because you have used it and you "know" it works, you have unit tests around it to ensure you exercise edge cases, etc.
```
Comment (iamalwaysuncomfortable) [inline attest/README.md:25]:
```suggestion
A TCB is created following good development practices. Stabilizing a procedure and including it in a library that is shared in multiple places makes that library part of a TCB. In doing so, ad-hoc trust in that procedure is established because it's been used extensively, tested to exercise edge cases, etc.
```

File: attest/README.md [hunk 15]
```diff
@@ -1,3 +1,108 @@
...
+
+All that testing and ad-hoc trust aside, there's an even deeper analogue to be learned. Ultimately, you want to be in the business of squashing *classes* of security bugs. It's all fine and good to fix a SQL injection bug or a string misquote, but what you really want to do is make your code mis-use resistant, so SQL injection or string misquoting bugs are simply impossible to write.
```
Comment (iamalwaysuncomfortable) [inline attest/README.md:27]:
```suggestion
Outside of informal trust, it is important that you're protecting against *classes* of security bugs. Code should be resistant to mis-use. For example, in a hypothetical system relying on SQL, SQL injection vulnerabilities shouldn't needed to be patched, such vulnerabilities should be impossible to write in the first place.
```

File: attest/README.md [hunk 16]
```diff
@@ -1,3 +1,108 @@
...
+
+Another way to get to the same place, which seems much weaker but actually has a richer *history* is to make bugs like those *not security impacting*. That is, OK, fine, so there's a SQL injection. If you are talking to a database with read-only credentials with restricted query times, that's not great, but it's also not that bad.
```
Comment (iamalwaysuncomfortable) [inline attest/README.md:29]:
```suggestion
```

**Review by iamalwaysuncomfortable — COMMENTED** [review]
Partial list of changes, will revisit later

File: account-keys/slip10/README.md [hunk 17]
```diff
@@ -1,5 +1,46 @@
-# MobileCoin SLIP-0010-Based Key Derivation
+# MobileCoin Key Derivation
 
-This crate provides utilities to handle SLIP-0010 key bytes and their relation to the MobileCoin [`AccountKey`](mc_account_keys::AccountKey) structure, which contains a pair of Ristretto255 view/spend private scalars.
+This crate implements a standard method for deriving a MobileCoin [`AccountKey`](mc_account_keys::AccountKey) from entropy using the following procedure:
 
-This also features a trait to create a Slip10Key from entropy and path, along with the canonical method of converting a BIP-39 [`Mnemonic`](tiny_bip32::Mnemonic) with a given BIP-32 path into a [`Slip10Key`](Slip10Key) usable within MobileCoin.
+1. A cryptographically secure random number generator is used to gather 32 bytes of entropy.
```
[the remark this answers is not in the document]
Comment (jcape) [reply account-keys/slip10/README.md:5]: We don't depend on operating system RNG, though, anything which implements the `CryptoRng` marker trait is accepted (e.g. the `McRng` object will use `RDRAND` inside an enclave).
